sklearn/manifold/tests/test_t_sne.py

```diff
@@ -872,13 +884,64 @@ def test_tsne_with_different_distance_metrics():
     for metric, dist_func in zip(metrics, dist_funcs):
         X_transformed_tsne = TSNE(
             metric=metric, n_components=n_components_embedding,
-            random_state=0, n_iter=300).fit_transform(X)
+            random_state=0, n_iter=300, square_distances=True).fit_transform(X)
         X_transformed_tsne_precomputed = TSNE(
             metric='precomputed', n_components=n_components_embedding,
-            random_state=0, n_iter=300).fit_transform(dist_func(X))
+            random_state=0, n_iter=300,
+            square_distances=True).fit_transform(dist_func(X))
         assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
 
 
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+@pytest.mark.parametrize('metric', ['euclidean', 'manhattan'])
+@pytest.mark.parametrize('square_distances', [True, 'legacy'])
+@ignore_warnings(category=FutureWarning)
+def test_tsne_different_square_distances(method, metric, square_distances):
+    # Make sure that TSNE works for different square_distances settings
+    # FIXME remove test when square_distances=True becomes the default in 0.26
+    random_state = check_random_state(0)
+    n_components_original = 3
+    n_components_embedding = 2
+
+    # Used to create data with structure; this avoids unstable behavior in TSNE
+    X, _ = make_blobs(n_features=n_components_original,
+                      random_state=random_state)
+    X_precomputed = pairwise_distances(X, metric=metric)
+
+    if metric == 'euclidean' and square_distances == 'legacy':
+        X_precomputed **= 2
```

Comment:
I'm a bit confused by this condition: shouldn't this be

```py
if metric == 'euclidean' or square_distances is True
```

?

does that mean that when we have (manhattan, True) we don't expect the distances to be squared? Or are the distances squared already when calling `X_precomputed = pairwise_distances(X, metric=metric)`?

Reply:
with `euclidean` + `legacy`, distances are squared so to be able to compare with precomputed we need to provide a squared distance matrix.

Reply:
But don't we expect the distances to be squared as well with `manhattan`, `True`?

Reply:
with True, distances are squared, whether it's a string metric (say `manhattan`) or a precomputed distance matrix (say `pairwise_distances(X, metric="manhattan")`)

Reply:
got it, thanks